### DISCUSSION TITLE
Suggestion: drop uncurried support of `create`

### DISCUSSION THREAD

**Discussion by devanshj** [pmndrs/zustand#1865] [Ideas]
I've lost count of how many people have missed the extra parenthesis in `create()(...)` (and the wrong usage ie `create(...)` doesn't even get caught because it's allowed). I don't see how supporting the `create(...)` usage adds any value, it's just creating trouble for many people (including me). I know I've suggested this before but I want to make the suggestion official hence opening an issue.

**Comment by dai-shi** [comment]
Thanks for opening a suggestion. Let's wait others to comment. One note is as such a change is breaking, it will be only possible in v5.

That said, my personal comment would be a) `create(...)` is still very valid and natural syntax for JS users, and b) `create()(...)` simply doesn't look comfortable, so I prefer to use `create(...)` without middleware, and I'm not a big fan of middleware. They are mostly preference, and my mind could be changed on this.

(I understand there are many troubles, and wonder if there's any other way to mitigate it, like warning for the usage of `create(...)` only in TS. Otherwise, `eslint-plugin-zustand` would be one solution. Or, hm, we could attach `api.middlwareIsUsed` property and warn it.)

**Comment by devanshj** [comment]
There's also an option to support it in runtime but not include it in types (ie .d.ts won't include the uncurried overload).

**Comment by Gustav-Eikaas** [comment]
Maybe i'm misunderstanding but is this the two types of syntaxes we're discussing?

```TS
type Test = {
  name: string;
};

const curried = create<Test>()((set) => ({ name: "Yes" }));

const uncurried = create<Test>((set) => ({
  name: "string",
}));
```

Personally I prefer the uncurried one and find the curried one weird..
Callback initializers are pretty common in the world of JS

**Comment by dai-shi** [comment]
> Maybe i'm misunderstanding but is this the two types of syntaxes we're discussing?

Yeah, in summary:
- In TypeScript, the curried one is _required_ if you use middleware (not all of them, but with store mutation, but hard to know which).
- The curried one works for all cases, both in JavaScript and TypeScript
- In JavaScript, the uncurried one works for all cases.

**Comment by ollwenjones** [comment]
We use TS, and have used the uncurried version almost exclusively, and have not had any problems, I guess because we aren't using middleware that requires it. We tend to use type inference on factory function return types for different slices, with good results.

**Comment by charkour** [comment]
The idea of `eslint-plugin-zustand` would be great. I've been wanting to make one so we can add optional rules like `always-use-selector` when calling `useStore` in a React component.

**Reply by dai-shi** [reply]
Seems like `eslint-plugin-zustand` would be a viable solution.

**Comment by dbritto-dev** [comment]
There's a way to support uncurried version for TS as well with middlewares? I would like to have just one way to do both, with and without middlewares.

**Comment by ollwenjones** [comment]
@dbritto-dev we aren't currying, but its fine because we aren't using middle-wares. It seems like some middle-wares require it.

**Comment by dbritto-dev** [comment]
> @dbritto-dev we aren't currying, but its fine because we aren't using middle-wares. It seems like some middle-wares require it.

@ollwenjones actually, I would like to have just one way to do both

**Comment by dai-shi** [comment]
https://github.com/pmndrs/zustand/pull/2137#issuecomment-1774224119

@devanshj I don't remember we discussed this before, but is there any possibility to avoid currying, if we introduce a new function like `applyMiddleware(...)`, hopefully without changing `createStore` JS code, or may be just a little bit?

**Reply by dai-shi** [reply]
well, it's unlikely...

(For anyone curious, here's the original detailed post: https://github.com/pmndrs/zustand/issues/710 )

**Reply by devanshj** [reply]
I don't understand what's `applyMiddleware(...)` but regardless there's no way to avoid the currying.

**Reply by dai-shi** [reply]
Thanks for your comment. Never mind about `applyMiddleware(...)`, which could be anything if it can avoid the currying.

Seems like eslint-plugin-zustand would be a viable solution. 👉  https://github.com/pmndrs/zustand/discussions/1865#discussioncomment-7375841

**Comment by dai-shi** [comment]
I would like to promote the use of `zustand-slices` with which works well with the uncurried version.

**Reply by dai-shi** [reply]
So, that makes me feel that dropping uncurried support is not a good idea.
